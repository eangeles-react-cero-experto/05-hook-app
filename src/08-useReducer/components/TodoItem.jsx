import PropTypes from "prop-types";

export const TodoItem = ({ todoItem }) => {
  return (
    <>
      <li className="list-group-item d-flex justify-content-between align-items-center">
        <span>{todoItem.description}</span>
        <button className="btn btn-sm btn-outline-danger">Remove</button>
      </li>
    </>
  );
};

TodoItem.propTypes = {
  todoItem: PropTypes.shape({
    id: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    isDone: PropTypes.bool.isRequired,
  }),
};
