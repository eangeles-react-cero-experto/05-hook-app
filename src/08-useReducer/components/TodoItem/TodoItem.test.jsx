import { fireEvent, render, screen } from "@testing-library/react";
import { TodoItem } from "./TodoItem.jsx";
import * as crypto from "crypto";

describe("Test in TodoItem", () => {
  let todoItem;
  const mockOnToggleTodoItem = jest.fn();
  const mockOnDeleteTodoItem = jest.fn();

  beforeEach(() => jest.clearAllMocks());
  beforeEach(() => {
    todoItem = {
      id: crypto.randomUUID(),
      description: "Walk the dog",
      isDone: false,
    };
  });

  it("should match with snapshot", () => {
    const { container } = render(
      <TodoItem
        todoItem={todoItem}
        onToggleTodoItem={mockOnToggleTodoItem}
        onDeleteTodoItem={mockOnDeleteTodoItem}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should show a uncompleted todo item without marked", () => {
    render(
      <TodoItem
        todoItem={todoItem}
        onToggleTodoItem={mockOnToggleTodoItem}
        onDeleteTodoItem={mockOnDeleteTodoItem}
      />,
    );

    const spanElement = screen.getByText(todoItem.description);
    expect(spanElement.classList).not.toContain("text-decoration-line-through");
  });

  it("should show a completed todo item as marked", () => {
    todoItem.isDone = true;
    render(
      <TodoItem
        todoItem={todoItem}
        onToggleTodoItem={mockOnToggleTodoItem}
        onDeleteTodoItem={mockOnDeleteTodoItem}
      />,
    );

    const spanElement = screen.getByText(todoItem.description);
    expect(spanElement.classList).toContain("text-decoration-line-through");
  });

  it("should call onToggleTodoItem when li is click", () => {
    render(
      <TodoItem
        todoItem={todoItem}
        onToggleTodoItem={mockOnToggleTodoItem}
        onDeleteTodoItem={mockOnDeleteTodoItem}
      />,
    );

    const liElement = screen.getByRole("listitem");
    fireEvent.click(liElement);

    expect(mockOnToggleTodoItem).toHaveBeenCalledTimes(1);
  });

  it("should call onDeleteTodoItem when remove button is click", () => {
    render(
      <TodoItem
        todoItem={todoItem}
        onToggleTodoItem={mockOnToggleTodoItem}
        onDeleteTodoItem={mockOnDeleteTodoItem}
      />,
    );

    const removeBtn = screen.getByRole("button", { name: "Remove" });
    fireEvent.click(removeBtn);

    expect(mockOnDeleteTodoItem).toHaveBeenCalledTimes(1);
  });
});
