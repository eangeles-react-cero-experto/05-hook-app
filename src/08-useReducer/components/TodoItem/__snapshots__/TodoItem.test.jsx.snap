// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in TodoItem should match with snapshot 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between align-items-center"
  >
    <span
      class="w-100 "
    >
      Walk the dog
    </span>
    <button
      class="btn btn-sm btn-outline-danger"
    >
      Remove
    </button>
  </li>
</div>
`;
